apiserver/paasng/paas_wl/cnative/specs/svc_disc.py

```diff
@@ -0,0 +1,133 @@
+"""
+TencentBlueKing is pleased to support the open source community by making
+蓝鲸智云 - PaaS 平台 (BlueKing - PaaS System) available.
+Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
+Licensed under the MIT License (the "License"); you may not use this file except
+in compliance with the License. You may obtain a copy of the License at
+
+    http://opensource.org/licenses/MIT
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+either express or implied. See the License for the specific language governing permissions and
+limitations under the License.
+
+We undertake not to change the open source license (MIT license) applicable
+to the current version of the project delivered to anyone in the future.
+"""
+import base64
+import json
+import logging
+from typing import Any, Dict, List
+
+from django.utils.encoding import force_bytes, force_str
+
+from paas_wl.cnative.specs.crd.bk_app import BkAppResource
+from paas_wl.platform.applications.models import WlApp
+from paas_wl.resources.base.exceptions import ResourceMissing
+from paas_wl.resources.base.kres import KConfigMap
+from paas_wl.resources.utils.basic import get_client_by_app
+from paasng.extensions.declarative.deployment.resources import BkSaaSItem
+from paasng.extensions.declarative.deployment.svc_disc import BkSaaSAddrDiscoverer
+from paasng.platform.applications.models import ModuleEnvironment
+
+logger = logging.getLogger(__name__)
+
+
+def apply_configmap(env: ModuleEnvironment, bk_app_res: BkAppResource):
+    """Apply the ConfigMap resource that contains the service discovery data to the
+    given environment, the ConfigMap might be deleted if the service discovery config
+    is absent.
+
+    :param env: The environment to apply the ConfigMap.
+    :param bk_app_res: The BkAppResource object.
+    """
+    svc_disc = bk_app_res.spec.svcDiscovery
+    mgr = ConfigMapManager(env, bk_app_name=bk_app_res.metadata.name)
+    if not (svc_disc and svc_disc.bkSaaS):
+        logger.debug('No service discovery config found, remove the ConfigMap if exists')
+        mgr.remove()
```

Comment:
apply_configmap通常会被理解为 update_or_create，但这里包含了删除逻辑(虽然通过函数注释进行了说明）。
- 建议换个名字(有点难想，还是因为这个函数做了很多事情)
- 不删除怎么样(只要保证不会和用户的configmap冲突)，可能大部分应用都不会用到svcDiscovery, 但每次都要remove configmap

Reply:
- 可以改成 sync 之类的名字，表示“同步”configmap资源，不过感觉差别不大
- 目前还是保留删吧，逻辑合理点，多一次调用影响倒不大，之后可以改成判断到“从有到无”时，才调用删除

不过倒是想起另一个事情。就是当应用重新部署时，调整或删除了其中服务发现的配置，那么那些已经挂载到现有应用 Pod 里的 json 文件内容，会在新 Pod 滚动前就马上发生变化。甚至假如说之后的发布出错了，那么 ConfigMap 的更新仍然那里（没有回滚）。

有个规避的方式，就是让应用开发者在读 json 文件时，忽略 SaaS 地址被删除的情况，地址总是只追加和更新。（不过这也解决不了 Pod 重新滚动的问题）。

我把这两个事项记录一下，第二个落地成单据。

```
        # TODO: Only remove the configmap if the app previously had a valid svc-discovery
        # config, don't perform the remove() operation every time.

        # NOTE: This action might break running pods that get svc-discovery data by
        # mounting the configmap as file, because some data might be removed in the
        # latest version. We should ask the application developer to handle this properly.
```